mechanics/tools/gdocs-migrate/forest.go

```diff
@@ -0,0 +1,63 @@
+package main
+
+type File struct {
+	id       string
+	name     string
+	parentID string
+	isFolder bool
+}
+
+type Forest struct {
+	fileByID map[string]File
+	children map[string][]File
+	leaves   []File
+}
+
+func NewForest() Forest {
+	return Forest{
+		fileByID: make(map[string]File),
+		children: make(map[string][]File),
+	}
+}
+
+func (fo *Forest) Add(f File) {
+	fo.fileByID[f.id] = f
+
+	fo.children[f.parentID] = append(fo.children[f.parentID], f)
+	if !f.isFolder {
+		fo.leaves = append(fo.leaves, f)
+	}
+}
+
+func (fo *Forest) GetFiles() []File {
+	return fo.leaves
+}
+
+func (fo *Forest) Size() int {
+	return len(fo.fileByID)
+}
+
+func (fo *Forest) FolderCount() int {
+	return fo.Size() - len(fo.GetFiles())
+}
+
+func (fo *Forest) GetPath(f File) []string {
+	parents := []string{}
+	cur := f
+	p, hasParent := fo.fileByID[cur.parentID]
+	for hasParent {
+		if p.name != "" {
+			parents = append(parents, p.name)
```

Comment:
Should this be a panic or something, rather than skipping over parents that don't have names?

Reply:
It's leftover -- I've updated the code.